crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1043,6 +1043,80 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
 
         Ok(())
     }
+
+
+/// Represents a single line in a docstring.
+///
+/// This type is used to both represent the original lines in a docstring
+/// (the line will be borrowed) and also the newly formatted lines from code
+/// snippets (the line will be owned).
+#[derive(Clone, Debug)]
+struct DocstringLine<'src> {
+    /// The actual text of the line, not including the line terminator.
+    line: Cow<'src, str>,
+    /// The offset into the source document which this line corresponds to.
+    offset: TextSize,
+    /// Whether this is the last line in a docstring or not. "Last" lines have
+    /// some special treatment when printing.
+    is_last: bool,
+}
+
+impl<'src> DocstringLine<'src> {
+    /// Return this line, but with the given function applied to the text of
+    /// the line.
+    fn map(self, mut map: impl FnMut(&str) -> String) -> DocstringLine<'static> {
+        DocstringLine {
+            line: Cow::Owned(map(&self.line)),
+            ..self
+        }
+    }
+}
+/// The kind of code example observed in a docstring.
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum CodeExampleKind {
+    /// Code found in Python "doctests."
+    ///
+    /// Documentation describing doctests and how they're recognized can be
+    /// found as part of the Python standard library:
+    /// https://docs.python.org/3/library/doctest.html.
+    ///
+    /// (You'll likely need to read the regex matching used internally by the
+    /// doctest module to determine more precisely how it works.)
```

Comment:
Could we add a link to where this regex can be found?

Reply:
There are some links further down in the actual doctest parsing code, but I've added another one here.